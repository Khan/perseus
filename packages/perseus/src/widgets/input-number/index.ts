export {default} from "./input-number";
