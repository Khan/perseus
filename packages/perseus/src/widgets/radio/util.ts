import * as i18n from "@khanacademy/wonder-blocks-i18n";

/**
 * Given a choice's position in the radio widget, return the corresponding
 * "choice letter". (For example, in English, `getChoiceLetter(0)` is "A",
 * `getChoiceLetter(1)` is "B", etc.)
 */
export function getChoiceLetter(pos: number): string {
    /* I18N: This is a list of single-character labels that will appear in
     * front of multiple-choice options. For instance, a multiple-choice
     * question with three options would display
     *  (A) first option
     *  (B) second option
     *  (C) third option
     * There must be spaces between each of the different characters. The
     * characters will show up next to options in the order that they are
     * listed here. Most multiple choice questions have 5 or fewer options.
     */
    const lettersString = i18n._(
        "A B C D E F G H I J K L M N O P Q R S T U V W X Y Z",
    );

    const letters = lettersString.split(" ");

    if (pos < letters.length) {
        // If the position we need is listed in the localized string, use that.
        return letters[pos];
    }
    // If we're out of letters, give up and return a space.
    return " ";
}
