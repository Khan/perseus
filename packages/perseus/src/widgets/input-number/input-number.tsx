import {linterContextDefault} from "@khanacademy/perseus-linter";
import {spacing} from "@khanacademy/wonder-blocks-tokens";
import {StyleSheet} from "aphrodite";
import * as React from "react";
import _ from "underscore";

import {PerseusI18nContext} from "../../components/i18n-context";
import InputWithExamples from "../../components/input-with-examples";
import SimpleKeypadInput from "../../components/simple-keypad-input";
import {ApiOptions} from "../../perseus-api";
import {getPromptJSON as _getPromptJSON} from "../../widget-ai-utils/input-number/input-number-ai-utils";

import scoreInputNumber, {answerTypes} from "./score-input-number";

import type {PerseusInputNumberWidgetOptions} from "../../perseus-types";
import type {PerseusStrings} from "../../strings";
import type {Path, Widget, WidgetExports, WidgetProps} from "../../types";
import type {
    PerseusInputNumberRubric,
    PerseusInputNumberUserInput,
} from "../../validation.types";
import type {InputNumberPromptJSON} from "../../widget-ai-utils/input-number/input-number-ai-utils";

const formExamples = {
    integer: function (options, strings: PerseusStrings) {
        return strings.integerExample;
    },
    proper: function (options, strings: PerseusStrings) {
        if (options.simplify === "optional") {
            return strings.properExample;
        }
        return strings.simplifiedProperExample;
    },
    improper: function (options, strings: PerseusStrings) {
        if (options.simplify === "optional") {
            return strings.improperExample;
        }
        return strings.simplifiedImproperExample;
    },
    mixed: function (options, strings: PerseusStrings) {
        return strings.mixedExample;
    },
    decimal: function (options, strings: PerseusStrings) {
        return strings.decimalExample;
    },
    percent: function (options, strings: PerseusStrings) {
        return strings.percentExample;
    },
    pi: function (options, strings: PerseusStrings) {
        return strings.piExample;
    },
} as const;

type RenderProps = {
    simplify: PerseusInputNumberWidgetOptions["simplify"];
    size: PerseusInputNumberWidgetOptions["size"];
    answerType: PerseusInputNumberWidgetOptions["answerType"];
    rightAlign: PerseusInputNumberWidgetOptions["rightAlign"];
};

type ExternalProps = WidgetProps<RenderProps, PerseusInputNumberRubric>;
type Props = ExternalProps & {
    apiOptions: NonNullable<ExternalProps["apiOptions"]>;
    linterContext: NonNullable<ExternalProps["linterContext"]>;
    rightAlign: NonNullable<ExternalProps["rightAlign"]>;
    size: NonNullable<ExternalProps["size"]>;
    currentValue: string;
    // NOTE(kevinb): This was the only default prop that is listed as
    // not-required in PerseusInputNumberWidgetOptions.
    answerType: NonNullable<PerseusInputNumberRubric["answerType"]>;
};

type DefaultProps = {
    answerType: Props["answerType"];
    apiOptions: Props["apiOptions"];
    currentValue: Props["currentValue"];
    linterContext: Props["linterContext"];
    rightAlign: Props["rightAlign"];
    size: Props["size"];
};

class InputNumber extends React.Component<Props> implements Widget {
    static contextType = PerseusI18nContext;
    declare context: React.ContextType<typeof PerseusI18nContext>;

    static defaultProps: DefaultProps = {
        currentValue: "",
        size: "normal",
        answerType: "number",
        rightAlign: false,
        // NOTE(kevinb): renderer.jsx should be provide this so we probably don't
        // need to include it in defaultProps.
        apiOptions: ApiOptions.defaults,
        linterContext: linterContextDefault,
    };

    static getUserInputFromProps(props: Props): {
        currentValue: string;
    } {
        return {
            currentValue: props.currentValue,
        };
    }

    shouldShowExamples: () => boolean = () => {
        return this.props.answerType !== "number";
    };

    handleChange: (arg1: string, arg2: () => void) => void = (newValue, cb) => {
        this.props.onChange({currentValue: newValue}, cb);
    };

    _handleFocus: () => void = () => {
        this.props.onFocus([]);
    };

    _handleBlur: () => void = () => {
        this.props.onBlur([]);
    };

    focus: () => boolean = () => {
        // eslint-disable-next-line react/no-string-refs
        // @ts-expect-error - TS2339 - Property 'focus' does not exist on type 'ReactInstance'.
        this.refs.input.focus();
        return true;
    };

    focusInputPath: (arg1: Path) => void = (inputPath) => {
        // eslint-disable-next-line react/no-string-refs
        // @ts-expect-error - TS2339 - Property 'focus' does not exist on type 'ReactInstance'.
        this.refs.input.focus();
    };

    blurInputPath: (arg1: Path) => void = (inputPath) => {
        // eslint-disable-next-line react/no-string-refs
        // @ts-expect-error - TS2339 - Property 'blur' does not exist on type 'ReactInstance'.
        if (typeof this.refs.input?.blur === "function") {
            // eslint-disable-next-line react/no-string-refs
            // @ts-expect-error - TS2339 - Property 'blur' does not exist on type 'ReactInstance'.
            this.refs.input?.blur();
        }
    };

    getInputPaths: () => ReadonlyArray<Path> = () => {
        // The widget itself is an input, so we return a single empty list to
        // indicate this.
        /* c8 ignore next */
        return [[]];
    };

    setInputValue: (arg1: Path, arg2: string, arg3: () => void) => void = (
        path,
        newValue,
        cb,
    ) => {
        this.props.onChange(
            {
                currentValue: newValue,
            },
            cb,
        );
    };

    getUserInput(): PerseusInputNumberUserInput {
        return InputNumber.getUserInputFromProps(this.props);
    }

    getPromptJSON(): InputNumberPromptJSON {
        return _getPromptJSON(this.props, this.getUserInput());
    }

    examples: () => ReadonlyArray<string> = () => {
        const {strings} = this.context;
        const type = this.props.answerType;
        const forms = answerTypes[type].forms.split(/\s*,\s*/);

        const examples = _.map(forms, (form) =>
            formExamples[form](this.props, strings),
        );

        return [strings.yourAnswer].concat(examples);
    };

    render(): React.ReactNode {
        if (this.props.apiOptions.customKeypad) {
            // TODO(charlie): Support "Review Mode".
            const input = (
                <SimpleKeypadInput
                    // eslint-disable-next-line react/no-string-refs
                    ref="input"
                    value={this.props.currentValue}
                    keypadElement={this.props.keypadElement}
                    onChange={this.handleChange}
                    onFocus={this._handleFocus}
                    onBlur={this._handleBlur}
                />
            );

            if (this.props.rightAlign) {
                return <div className="perseus-input-right-align">{input}</div>;
            }

            return input;
        }

        // Note: This is _very_ similar to what `numeric-input.jsx` does. If
        // you modify this, double-check if you also need to modify that
        // component.
        const inputStyles = [
            styles.default,
            this.props.size === "small" ? styles.small : null,
            this.props.rightAlign ? styles.rightAlign : styles.leftAlign,
        ];
        // Unanswered
        if (this.props.reviewMode && !this.props.currentValue) {
            inputStyles.push(styles.answerStateUnanswered);
        }

        return (
            <InputWithExamples
                // eslint-disable-next-line react/no-string-refs
                ref="input"
                value={this.props.currentValue}
                onChange={this.handleChange}
                style={inputStyles}
                examples={this.examples()}
                shouldShowExamples={this.shouldShowExamples()}
                onFocus={this._handleFocus}
                onBlur={this._handleBlur}
                id={this.props.widgetId}
                disabled={this.props.apiOptions.readOnly}
                linterContext={this.props.linterContext}
            />
        );
    }
}

const styles = StyleSheet.create({
    default: {
        width: 80,
        height: "auto",
    },
    small: {
        width: 40,
    },
    leftAlign: {
        paddingLeft: spacing.xxxSmall_4,
        paddingRight: 0,
    },
    rightAlign: {
        textAlign: "right",
        paddingLeft: 0,
        paddingRight: spacing.xxxSmall_4,
    },
    answerStateUnanswered: {
        backgroundColor: "#eee",
        border: "solid 1px #999",
    },
});

const propTransform = (
    widgetOptions: PerseusInputNumberWidgetOptions,
): RenderProps => {
    const {simplify, size, answerType, rightAlign} = widgetOptions;
    return {
        simplify,
        size,
        answerType,
        rightAlign,
    };
};

export default {
    name: "input-number",
    displayName: "Input number (deprecated - use numeric input instead)",
    defaultAlignment: "inline-block",
    hidden: true,
    widget: InputNumber,
    transform: propTransform,
    isLintable: true,
    scorer: scoreInputNumber,

    getOneCorrectAnswerFromRubric(rubric: any): string | null | undefined {
        if (rubric.value == null) {
            return;
        }
        let answerString = String(rubric.value);
        if (rubric.inexact && rubric.maxError) {
            answerString += " \u00B1 " + rubric.maxError;
        }
        return answerString;
    },
} satisfies WidgetExports<typeof InputNumber>;
