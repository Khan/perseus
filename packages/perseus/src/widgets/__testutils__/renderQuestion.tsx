import {RenderStateRoot} from "@khanacademy/wonder-blocks-core";
import {render} from "@testing-library/react";
import * as React from "react";

import {
    testDependenciesV2,
    testDependencies,
    // eslint-disable-next-line import/no-relative-packages
} from "../../../../../testing/test-dependencies";
import {
    DependenciesContext,
    useDependencies,
    setDependencies,
} from "../../dependencies";
import * as Perseus from "../../index";
import {mockStrings} from "../../strings";
import {registerAllWidgetsForTesting} from "../../util/register-all-widgets-for-testing";

import type {APIOptions} from "../../types";
import type {PerseusRenderer} from "@khanacademy/perseus-core";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

type RenderResult = ReturnType<typeof render>;

type ExtraProps = Omit<PropsFor<typeof Perseus.Renderer>, "strings">;

export const renderQuestion = (
    question: PerseusRenderer,
    apiOptions: APIOptions = Object.freeze({}),
    extraProps?: ExtraProps,
): {
    container: HTMLElement;
    renderer: Perseus.Renderer;
    rerender: (question: PerseusRenderer, extraProps?: ExtraProps) => void;
    unmount: RenderResult["unmount"];
} => {
    setDependencies(testDependencies);
    registerAllWidgetsForTesting();

    let renderer: Perseus.Renderer | null = null;
    const {container, rerender, unmount} = render(
        <RenderStateRoot>
            <DependenciesContext.Provider value={testDependenciesV2}>
                <Renderer
                    ref={(node) => (renderer = node)}
                    question={question as any}
                    apiOptions={apiOptions}
                    extraProps={{
                        ...extraProps,
                        strings: mockStrings,
                    }}
                />
            </DependenciesContext.Provider>
        </RenderStateRoot>,
    );
    // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
    if (!renderer) {
        throw new Error(`Failed to render!`);
    }
    const renderAgain = (
        question: PerseusRenderer,
        extraProps?: ExtraProps,
    ) => {
        rerender(
            <RenderStateRoot>
                <DependenciesContext.Provider value={testDependenciesV2}>
                    <Renderer
                        ref={(node) => (renderer = node)}
                        question={question}
                        apiOptions={apiOptions}
                        extraProps={{
                            ...extraProps,
                            strings: mockStrings,
                        }}
                    />
                </DependenciesContext.Provider>
            </RenderStateRoot>,
        );
        if (!renderer) {
            throw new Error(`Failed to rerender!`);
        }
    };

    return {container, renderer, rerender: renderAgain, unmount};
};

const Renderer = React.forwardRef<
    Perseus.Renderer,
    {
        question: PerseusRenderer;
        apiOptions: APIOptions;
        extraProps?: PropsFor<typeof Perseus.Renderer>;
    }
>((props, ref) => {
    const dependencies = useDependencies();
    return (
        <Perseus.Renderer
            ref={ref}
            content={props.question.content}
            images={props.question.images}
            widgets={props.question.widgets}
            problemNum={0}
            apiOptions={props.apiOptions}
            strings={mockStrings}
            {...props.extraProps}
            {...dependencies}
        />
    );
});
