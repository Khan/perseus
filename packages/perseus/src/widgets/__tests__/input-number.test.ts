/**
 * Disclaimer: Definitely not thorough enough
 */
import {describe, beforeEach, it} from "@jest/globals";
import {screen} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import _ from "underscore";

import {testDependencies} from "../../../../../testing/test-dependencies";
import * as Dependencies from "../../dependencies";
import {errors} from "../../util/answer-types";
import {question3 as question} from "../__testdata__/input-number.testdata";
import InputNumber from "../input-number";

import {renderQuestion} from "./renderQuestion";

import type {
    PerseusInputNumberWidgetOptions,
    PerseusRenderer,
} from "../../perseus-types";

const {transform} = InputNumber;

const options: PerseusInputNumberWidgetOptions = {
    value: "2^{-2}-3",
    size: "normal",
    simplify: "optional",
};

describe("input-number", function () {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    describe("full render", function () {
        it("Shoud accept the right answer", () => {
            // Arrange
            const {renderer} = renderQuestion(question);

            // Act
            userEvent.paste(screen.getByRole("textbox"), "1/2");

            // Assert
            expect(renderer).toHaveBeenAnsweredCorrectly();
        });

        it("should reject an incorrect answer", () => {
            // Arrange
            const {renderer} = renderQuestion(question);

            // Act
            userEvent.paste(screen.getByRole("textbox"), "0.7");

            // Assert
            expect(renderer).toHaveBeenAnsweredIncorrectly();
        });

        it("should refuse to score an incoherent answer", () => {
            // Arrange
            const {renderer} = renderQuestion(question);

            // Act
            userEvent.paste(screen.getByRole("textbox"), "0..7");

            // Assert
            expect(renderer).toHaveInvalidInput();
        });
    });

    describe.each([
        [
            {
                content:
                    "Denis baked a peach pie and cut it into $3$ equal-sized pieces.  Denis's dad eats $1$ section of the pie.  \n\n**What fraction of the pie did Denis's dad eat?**  \n![](https://ka-perseus-graphie.s3.amazonaws.com/74a2b7583a2c26ebfb3ad714e29867541253fc97.png)    \n[[\u2603 input-number 1]]  \n\n\n\n",
                images: {
                    "https://ka-perseus-graphie.s3.amazonaws.com/74a2b7583a2c26ebfb3ad714e29867541253fc97.png":
                        {
                            width: 200,
                            height: 200,
                        },
                },
                widgets: {
                    "input-number 1": {
                        version: {
                            major: 0,
                            minor: 0,
                        },
                        type: "input-number",
                        graded: true,
                        alignment: "default",
                        options: {
                            maxError: 0.1,
                            inexact: false,
                            value: 0.3333333333333333,
                            simplify: "optional",
                            answerType: "rational",
                            size: "normal",
                        },
                    },
                },
            } as PerseusRenderer,
            "1/3",
            "0.4",
        ],
        [
            {
                content:
                    "Denis baked a peach pie and cut it into $3$ equal-sized pieces.  Denis's dad eats $1$ section of the pie.  \n\n**What fraction of the pie did Denis's dad eat?**  \n![](https://ka-perseus-graphie.s3.amazonaws.com/74a2b7583a2c26ebfb3ad714e29867541253fc97.png)    \n[[\u2603 input-number 1]]  \n\n\n\n",
                images: {
                    "https://ka-perseus-graphie.s3.amazonaws.com/74a2b7583a2c26ebfb3ad714e29867541253fc97.png":
                        {
                            width: 200,
                            height: 200,
                        },
                },
                widgets: {
                    "input-number 1": {
                        version: {
                            major: 0,
                            minor: 0,
                        },
                        type: "input-number",
                        graded: true,
                        alignment: "default",
                        options: {
                            maxError: 0.1,
                            inexact: false,
                            value: 0.3333333333333333,
                            simplify: "required",
                            answerType: "rational",
                            size: "normal",
                        },
                    },
                },
            } as PerseusRenderer,
            "1/3",
            "0.4",
        ],
        [
            {
                content:
                    "A washing machine is being redesigned to handle a greater volume of water.  One part is a pipe with a radius of $3 \\,\\text{cm}$ and a length of $11\\,\\text{cm}$.  It gets replaced with a pipe of radius $4\\,\\text{cm}$, and the same length. \n\n**How many more cubic centimeters of water can the new pipe hold?**\n\n [[\u2603 input-number 1]] $\\text{cm}^3$",
                images: {},
                widgets: {
                    "input-number 1": {
                        type: "input-number",
                        graded: true,
                        options: {
                            maxError: 0.1,
                            inexact: false,
                            value: 241.90263432641407,
                            simplify: "required",
                            answerType: "pi",
                            size: "normal",
                        },
                    },
                },
            } as PerseusRenderer,
            "77 pi",
            "76 pi",
        ],
        [
            {
                content:
                    'Akshat works in a hospital lab.\n\nTo project blood quantities, he wants to know the probability that more than $1$ of the next $7$ donors will have type-A blood. From his previous work, Sorin knows that $\\dfrac14$ of donors have type-A blood.\n\nAkshat uses a computer to produce many samples that simulate the next $7$ donors. The first $8$ samples are shown in the table below where "$\\text{\\red{A}}$" represents a donor *with* type-A blood, and "$\\text{\\blue{Z}}$" represents a donor *without* type-A blood.\n\n**Based on the samples below, estimate the probability that  more than $1$ of the next $7$ donors will have type-A blood.** If necessary, round your answer to the nearest hundredth. [[\u2603 input-number 1]]\n\n*Note: This a small sample to practice with. A larger sample could give a much better estimate.*\n\n | Sample |\n:-: | :-: | \n$1$ | $\\text{\\blue{Z}, \\blue{Z}, \\blue{Z}, \\blue{Z}, \\red{A}, \\blue{Z}, \\blue{Z}}$\n$2$ | $\\text{\\blue{Z}, \\blue{Z}, \\blue{Z}, \\blue{Z}, \\blue{Z}, \\blue{Z}, \\blue{Z}}$\n$3$ | $\\text{\\blue{Z}, \\blue{Z}, \\red{A}, \\blue{Z}, \\blue{Z}, \\blue{Z}, \\blue{Z}}$\n$4$ | $\\text{\\red{A}, \\red{A}, \\blue{Z}, \\blue{Z}, \\blue{Z}, \\blue{Z}, \\blue{Z}}$\n$5$ | $\\text{\\blue{Z}, \\blue{Z}, \\blue{Z}, \\blue{Z}, \\blue{Z}, \\red{A}, \\red{A}}$\n$6$ | $\\text{\\blue{Z}, \\red{A}, \\red{A}, \\blue{Z}, \\blue{Z}, \\blue{Z}, \\blue{Z}}$\n$7$ | $\\text{\\blue{Z}, \\red{A}, \\blue{Z}, \\blue{Z}, \\blue{Z}, \\red{A}, \\blue{Z}}$\n$8$ | $\\text{\\blue{Z}, \\blue{Z}, \\blue{Z}, \\blue{Z}, \\red{A}, \\blue{Z}, \\blue{Z}}$\n\n',
                images: {},
                widgets: {
                    "input-number 1": {
                        type: "input-number",
                        graded: true,
                        options: {
                            maxError: 0.1,
                            inexact: false,
                            value: 0.5,
                            simplify: "optional",
                            answerType: "percent",
                            size: "small",
                        },
                    },
                },
            } as PerseusRenderer,
            "50%",
            "0.56",
        ],
    ])("answer type", (question, correct, incorrect) => {
        it("Shoud accept the right answer", () => {
            // Arrange
            const {renderer} = renderQuestion(question);

            // Act
            userEvent.paste(screen.getByRole("textbox"), correct);

            // Assert
            expect(renderer).toHaveBeenAnsweredCorrectly();
        });

        it("should reject an incorrect answer", () => {
            // Arrange
            const {renderer} = renderQuestion(question);

            // Act
            userEvent.paste(screen.getByRole("textbox"), incorrect);

            // Assert
            expect(renderer).toHaveBeenAnsweredIncorrectly();
        });
    });

    it("transform should remove the `value` field", function () {
        const editorProps = {
            value: 5,
            simplify: "required",
            size: "normal",
            inexact: false,
            maxError: 0.1,
            answerType: "number",
        } as const;
        if (!transform) {
            throw new Error("transform not defined");
        }
        const widgetProps = transform(editorProps);
        expect(_.has(widgetProps, "value")).toBe(false);
    });
});

describe("invalid", function () {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("should handle invalid answers with no error callback", function () {
        const err = InputNumber.widget.validate({currentValue: "x+1"}, options);
        expect(err).toStrictEqual({
            message: errors.EXTRA_SYMBOLS_ERROR,
            type: "invalid",
        });
    });
});

describe("getOneCorrectAnswerFromRubric", () => {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("should return undefined if rubric.value is null/undefined", () => {
        // Arrange
        const rubric: Record<string, any> = {};

        // Act
        const result = InputNumber.widget.getOneCorrectAnswerFromRubric(rubric);

        // Assert
        expect(result).toBeUndefined();
    });

    it("should return rubric.value if inexact is false", () => {
        // Arrange
        const rubric = {
            value: 0,
            maxError: 0.1,
            inexact: false,
        } as const;

        // Act
        const result = InputNumber.widget.getOneCorrectAnswerFromRubric(rubric);

        // Assert
        expect(result).toEqual("0");
    });

    it("should return rubric.value with an error band if inexact is true", () => {
        // Arrange
        const rubric = {
            value: 0,
            maxError: 0.1,
            inexact: true,
        } as const;

        // Act
        const result = InputNumber.widget.getOneCorrectAnswerFromRubric(rubric);

        // Assert
        expect(result).toEqual("0 ± 0.1");
    });
});

describe("rendering", () => {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("supports mobile rendering", () => {
        const {container} = renderQuestion(question, {
            // Setting this triggers mobile rendering
            // it would be nice if this was more clear in the code
            customKeypad: true,
        });

        expect(container).toMatchSnapshot("mobile render");
    });
});

describe("focus state", () => {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("supports focusing", () => {
        //  Arrange
        const {renderer} = renderQuestion(question);

        // Act
        const gotFocus = renderer.focus();

        // Assert
        expect(gotFocus).toBeTrue();
    });

    it("supports blurring", () => {
        //  Arrange
        const {renderer} = renderQuestion(question);

        // Act
        const gotFocus = renderer.focus();
        renderer.blur();

        // Assert
        expect(gotFocus).toBeTrue();
    });

    it("calls the interaction callback on focus", () => {
        const testCallback = jest.fn();

        const {renderer} = renderQuestion(question, {
            interactionCallback: testCallback,
        });

        // Act
        renderer.focus();

        // Assert
        expect(testCallback).toHaveBeenCalled();
    });
});
