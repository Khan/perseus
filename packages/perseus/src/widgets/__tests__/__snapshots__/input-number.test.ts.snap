// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering supports mobile rendering: mobile render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        Akshat works in a hospital lab.
      </div>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="1"
    >
      <div
        class="paragraph"
      >
        To project blood quantities, he wants to know the probability that more than 
        <span
          style="white-space: nowrap;"
        >
          <span />
          <span
            class="mock-TeX"
          >
            1
          </span>
          <span />
        </span>
         of the next 
        <span
          style="white-space: nowrap;"
        >
          <span />
          <span
            class="mock-TeX"
          >
            7
          </span>
          <span />
        </span>
         donors will have type-A blood. From his previous work, Sorin knows that 
        <span
          style="white-space: nowrap;"
        >
          <span />
          <span
            class="mock-TeX"
          >
            \\dfrac14
          </span>
          <span />
        </span>
         of donors have type-A blood.
      </div>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="2"
    >
      <div
        class="paragraph"
      >
        Akshat uses a computer to produce many samples that simulate the next 
        <span
          style="white-space: nowrap;"
        >
          <span />
          <span
            class="mock-TeX"
          >
            7
          </span>
          <span />
        </span>
         donors. The first 
        <span
          style="white-space: nowrap;"
        >
          <span />
          <span
            class="mock-TeX"
          >
            8
          </span>
          <span />
        </span>
         samples are shown in the table below where "
        <span
          style="white-space: nowrap;"
        >
          <span />
          <span
            class="mock-TeX"
          >
            \\text{\\red{A}}
          </span>
          <span />
        </span>
        " represents a donor 
        <em>
          with
        </em>
         type-A blood, and "
        <span
          style="white-space: nowrap;"
        >
          <span />
          <span
            class="mock-TeX"
          >
            \\text{\\blue{Z}}
          </span>
          <span />
        </span>
        " represents a donor 
        <em>
          without
        </em>
         type-A blood.
      </div>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="3"
    >
      <div
        class="paragraph"
      >
        <strong>
          Based on the samples below, estimate the probability that  more than 
          <span
            style="white-space: nowrap;"
          >
            <span />
            <span
              class="mock-TeX"
            >
              1
            </span>
            <span />
          </span>
           of the next 
          <span
            style="white-space: nowrap;"
          >
            <span />
            <span
              class="mock-TeX"
            >
              7
            </span>
            <span />
          </span>
           donors will have type-A blood.
        </strong>
         If necessary, round your answer to the nearest hundredth. 
        <div
          class="perseus-widget-container widget-nohighlight widget-inline-block"
        >
          <div
            aria-label="Math input box Tap with one or two fingers to open keyboard"
            class="initial_4qg14c-o_O-input_1n1c032"
            role="textbox"
          >
            <div
              class="keypad-input"
              tabindex="0"
            >
              <div
                class="mq-editable-field mq-math-mode"
                style="background-color: white; min-height: 44px; min-width: 64px; max-width: 128px; box-sizing: border-box; position: relative; border-style: solid; border-color: rgba(33,36,44,0.50); border-radius: 4px; color: rgb(33, 36, 44); border-width: 1px;"
              >
                <span
                  class="mq-textarea"
                >
                  <span />
                </span>
                <span
                  class="mq-root-block mq-empty"
                  mathquill-block-id="1"
                  style="padding: 10px 11px 8px 11px; font-size: 18pt;"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="4"
    >
      <div
        class="paragraph"
      >
        <em>
          Note: This a small sample to practice with. A larger sample could give a much better estimate.
        </em>
      </div>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="5"
    >
      <table>
        <thead>
          <tr>
            <th
              scope="col"
              style="text-align: center;"
            />
            <th
              scope="col"
              style="text-align: center;"
            >
              Sample
            </th>
            <th
              scope="col"
            />
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              style="text-align: center;"
            >
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  1
                </span>
                <span />
              </span>
            </td>
            <td
              style="text-align: center;"
            >
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  \\text{\\blue{Z}, \\blue{Z}, \\blue{Z}, \\blue{Z}, \\red{A}, \\blue{Z}, \\blue{Z}}
                </span>
                <span />
              </span>
            </td>
          </tr>
          <tr>
            <td
              style="text-align: center;"
            >
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  2
                </span>
                <span />
              </span>
            </td>
            <td
              style="text-align: center;"
            >
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  \\text{\\blue{Z}, \\blue{Z}, \\blue{Z}, \\blue{Z}, \\blue{Z}, \\blue{Z}, \\blue{Z}}
                </span>
                <span />
              </span>
            </td>
          </tr>
          <tr>
            <td
              style="text-align: center;"
            >
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  3
                </span>
                <span />
              </span>
            </td>
            <td
              style="text-align: center;"
            >
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  \\text{\\blue{Z}, \\blue{Z}, \\red{A}, \\blue{Z}, \\blue{Z}, \\blue{Z}, \\blue{Z}}
                </span>
                <span />
              </span>
            </td>
          </tr>
          <tr>
            <td
              style="text-align: center;"
            >
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  4
                </span>
                <span />
              </span>
            </td>
            <td
              style="text-align: center;"
            >
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  \\text{\\red{A}, \\red{A}, \\blue{Z}, \\blue{Z}, \\blue{Z}, \\blue{Z}, \\blue{Z}}
                </span>
                <span />
              </span>
            </td>
          </tr>
          <tr>
            <td
              style="text-align: center;"
            >
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  5
                </span>
                <span />
              </span>
            </td>
            <td
              style="text-align: center;"
            >
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  \\text{\\blue{Z}, \\blue{Z}, \\blue{Z}, \\blue{Z}, \\blue{Z}, \\red{A}, \\red{A}}
                </span>
                <span />
              </span>
            </td>
          </tr>
          <tr>
            <td
              style="text-align: center;"
            >
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  6
                </span>
                <span />
              </span>
            </td>
            <td
              style="text-align: center;"
            >
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  \\text{\\blue{Z}, \\red{A}, \\red{A}, \\blue{Z}, \\blue{Z}, \\blue{Z}, \\blue{Z}}
                </span>
                <span />
              </span>
            </td>
          </tr>
          <tr>
            <td
              style="text-align: center;"
            >
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  7
                </span>
                <span />
              </span>
            </td>
            <td
              style="text-align: center;"
            >
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  \\text{\\blue{Z}, \\red{A}, \\blue{Z}, \\blue{Z}, \\blue{Z}, \\red{A}, \\blue{Z}}
                </span>
                <span />
              </span>
            </td>
          </tr>
          <tr>
            <td
              style="text-align: center;"
            >
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  8
                </span>
                <span />
              </span>
            </td>
            <td
              style="text-align: center;"
            >
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  \\text{\\blue{Z}, \\blue{Z}, \\blue{Z}, \\blue{Z}, \\red{A}, \\blue{Z}, \\blue{Z}}
                </span>
                <span />
              </span>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
